// soundworks/server/core/serviceManager.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _Signal = require('../../utils/Signal');

var _Signal2 = _interopRequireDefault(_Signal);

var _SignalAll = require('../../utils/SignalAll');

var _SignalAll2 = _interopRequireDefault(_SignalAll);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _ctors = {};
var _instances = {};

/**
 * Manager the services and their relations. Acts as a factory to ensure services
 * are instanciated only once.
 */
var serviceManager = {
  // add an init method
  /**
   *
   *
   */
  init: function init() {
    this._ready = this._ready.bind(this);

    this.signals = {};
    this.signals.start = new _Signal2.default();
    this.signals.ready = new _Signal2.default();

    this._requiredSignals = new _SignalAll2.default();
    this._requiredSignals.addObserver(this._ready);
  },
  start: function start() {
    this.signals.start.set(true);

    if (this._requiredSignals.length === 0) this._ready();
  },
  _ready: function _ready() {
    this.signals.ready.set(true);
  },


  /**
   * Retrieve a service according to the given id. If the service as not beeen
   * requested yet, it is instanciated.
   * @param {String} id - The id of the registered service
   * @param {Object} options - The options to configure the service.
   */
  require: function require(id) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    id = 'service:' + id;

    if (!_ctors[id]) throw new Error('Service "' + id + '" is not defined');

    var instance = _instances[id];

    if (!instance) {
      instance = new _ctors[id]();
      _instances[id] = instance;
    }

    instance.configure(options);
    return instance;
  },


  /**
   * Regiter a service
   * @param {String} id - The id of the service, in order to retrieve it later.
   * @param {Function} ctor - The constructor of the service.
   */
  register: function register(id, ctor) {
    _ctors[id] = ctor;
  },
  getRequiredServices: function getRequiredServices(clientType) {
    var services = [];

    for (var id in _instances) {
      if (_instances[id].clientTypes.has(clientType)) services.push(id);
    }

    return services;
  },
  getServiceList: function getServiceList() {
    return (0, _keys2.default)(_ctors);
  }
};

exports.default = serviceManager;
